Introduce dataframe to ExperimentData (step2)

nkanazawa1989 opened the pull request:
### Summary

Executing approved design proposal in https://github.com/Qiskit/rfcs/blob/master/0007-experiment-dataframe.md. This PR replaces the representation of curve data points with data frame. This object will be added to artifact in a follow up PR.

### Details and comments

In this PR, representation of the intermediate data for `CurveAnalysis` is replaced with the data frame `ScatterTable`. Experimentalists want easier access to the XY data points after curve analysis with various motivations, e.g. plotting data with their own code, rerunning analysis outside the CurveAnalysis framework, or analyzing the time series of curve data. 

A curve data consists of not only x, y values, but also multiple metadata such as associated fit model, process status (raw, formatted, fitted), and some circuit metadata per each data point. The data frame representation is convenient to manage such complicated data set, and also allows us to record all information in a single object.

In addition, `CurveAnalysis` gains `_create_figures` method thanks to `ScatterTable` representation, which cannot be implemented with conventional `CurveData` object. This allows a curve analysis subclass to overwrites the method to flexibly customize figure generation. For example, current `StarkRamseyXYAmpScanAnalysis` [overwrites entire `_run_analysis` method](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/c01b0fad86a42ffb3437757a146b79d501992cf4/qiskit_experiments/library/characterization/analysis/ramsey_xy_analysis.py#L539-L685) just to add second axis to the figure.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -314,36 +317,29 @@ def _run_curve_fit(
+        # Create convenient function to compute residual of the models.
+        partial_residuals = []
+        valid_uncertainty = np.all(np.isfinite(curve_data.yerr.to_numpy()))
+        for i, sub_data in list(curve_data.groupby("model_id")):
```
Filtering for each subset is now implemented by dataframe, but this creates another dataframe for subset. Running this in the objective function which is called many times by the optimizer degrades performance. Instead, this logic filters subset in advance and create partial-bound functions that evaluate the residual. This alleviates the performance bottleneck.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -380,53 +379,75 @@ def _objective(_params):
+        # Prepare for fitting
+        self._initialize(experiment_data)
+
+        curve_data = self._format_data(self._run_data_processing(experiment_data.data()))
```
This curve_data object is now `ScatterTable`. This table can store both processed_data and formatted_data and code becomes much simpler.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -441,58 +462,67 @@ def _run_analysis(
+                model_fit[:, 4] = i
+                # type
+                model_fit[:, 6] = "fit"
+            curve_data = curve_data.append_list_values(
```
This curve_data also stores fit values. Note that we can regenerate figure object from this data, i.e. lazy figure gen (but metadata such as chisq and fit parameters must be separately provided). In follow-up maybe we can update internals of curve plotter to directly take this object.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -135,122 +148,115 @@ def models(self) -> List[lmfit.Model]:
         raw_data: List[Dict],
-        models: List[lmfit.Model],
-    ) -> CurveData:
+    ) -> ScatterTable:
```
Is it okey to drop argument?

coruscating replied:
I'm okay with this since probably not many people were directly interacting with it.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -280,54 +332,31 @@ def _run_analysis(
     ) -> Tuple[List[AnalysisResultData], List["matplotlib.figure.Figure"]]:
 
         analysis_results = []
+        figures = []
```
This is almost copy of `CurveAnalysis` class, but will be completely dropped once after we move `ScatterTable` to artifact. Eventually I want to write

```python
component_analysis_results = []
for analysis in self._analyses:
    analysis.set_options(plot=False)
    sub_results, _ = analysis._run_analysis(experiment_data)
    component_analysis_results.extend(sub_results)

# --- I need this logic ---
# 1. get ScatterData of the sub analysis from the artifact
# 2. combine sub scatter data
figures = self._create_figures(combined_scatter_data)

return component_analysis_results + composite_results
```

but currently scatter data is discarded. Note that composite curve analysis usually creates a single plot including curves from sub analysis, c.f. HamiltonianTomo. Current implementation is just to keep scatter data of the sub analyses, so do not need to review carefully.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/utils.py:
```diff
@@ -222,6 +224,166 @@ def eval_with_uncertainties(
     return wrapfunc(x=x, **sub_params)
 
 
+def shot_weighted_average(
```
These functions are tied to ScatterTable column structure. I cannot reuse existing ones.

nkanazawa1989 reviewed (commented):
Some clue for review.

coruscating commented on requirements.txt:
```diff
@@ -5,4 +5,5 @@ qiskit-ibm-experiment>=0.3.4
-rustworkx
\ No newline at end of file
+rustworkx
+pandas>=1.1.5
```
Now that pandas is a requirement, the `requires_pandas()` decorator can be removed and so can the install pandas note in `tutorials/calibrations.rst`.

nkanazawa1989 replied:
Seems like the decorator is not used in the rst file, but text is updated and the decorator is removed from the entire package. [46d350e](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1253/commits/46d350e224823761b3711d24794e77bd5025b15b)

coruscating reviewed (commented):
Thanks @nkanazawa1989, this looks good overall, and modularizing the create figures functionality is nice.

The `CurveData` mentions in the Curve Analysis tutorial also need to be updated. As for the `auto_save` bug mentioned in my last comment, I think it's due to `AnalysisResultTable.add_entry()` returning the results of `DefaultColumnsMixIn.add_entry()`, which has no return value (missing a return?).

nkanazawa1989 commented on qiskit_experiments/curve_analysis/utils.py:
```diff
@@ -222,6 +224,166 @@ def eval_with_uncertainties(
+
+    if len(values) == 1:
+        out = values[0]
+        out[6] = "fit-ready"
```
This is reasonable suggestion. Done in [adec962](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1253/commits/adec962a5b758d60b9f83361a51682acdc51efd0)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -337,65 +366,78 @@ def _run_analysis(
+                    # model_id
+                    model_fit[:, 4] = i
+                    # type
+                    model_fit[:, 6] = "fit"
```
I changed all labels in [15a0f04](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1253/commits/15a0f045ec05d142d16fe8a02b0b29994f7f363d). 

- raw -> processed
- fit-ready -> formatted
- fit -> fitted

I feel these are more unified.

nkanazawa1989 commented:
Thanks @coruscating for detailed review and testing. The problem of auto saving was due to previous PR of https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1252, in which I removed the return value because returning the added `Series` actually creates new object from the table. This could be a performance bottleneck. The problem is fixed in [7b07559](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1253/commits/7b07559dbd058005578990a2e61c54f7874f31f4).

I also cleaned up the scatter table columns, namely, I dropped a builtin column of `formatted` because this is almost overlapped with table index. This slightly reduces memory footprint; [5c5b3b6](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1253/commits/5c5b3b62971bc86f722a489d57390838fcad8ad0).

I'm still struggling with removing superclass (DataFrame) members from the docstring. I guess I need to write a custom Sphinx directive that is counterpart of `:exclude-member:`, or maybe another directive that only renders members of current class.

coruscating reviewed (approved):
Thanks for the performance improvement, this looks good to me now!
